### PR TITLE
Syntax to control typeclass resolution db for rewrite_strat

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix https://github.com/coq/coq/issues/15093

### CODE REVIEW

**Comment by SkySkimmer** [comment]
@Janno @gmalecha please provide some examples to test this with

File: proofs/clenv.mli [hunk 1]
```diff
@@ -92,7 +92,7 @@ val clenv_push_prod : clausenv -> (metavariable * bool * clausenv) option
 (** {6 Clenv tactics} *)
 
 val unify : ?flags:unify_flags -> constr -> unit Proofview.tactic
-val res_pf : ?with_evars:bool -> ?with_classes:bool -> ?flags:unify_flags -> clausenv -> unit Proofview.tactic
+val res_pf : ?with_evars:bool -> ?with_classes:bool -> ?flags:unify_flags -> db:string -> clausenv -> unit Proofview.tactic
```
Comment (ppedrot) [inline proofs/clenv.mli:95]: Rather pass this as an option using the `with_classes` argument.

**Comment by SkySkimmer** [comment]
This is only for testing, don't bother reviewing the code.

**Comment by Janno** [comment]
Sorry to keep you waiting so long and thank you for keeping this PR updated! 

I just tried to evaluate this change against the test case in https://github.com/coq/coq/issues/16471. The original problem in that issue was one I found in an actual proof and it led to significant slowdown so I was hoping that it could be avoided by using `rewrite_strat .. with <empty_database>` to make sure `rewrite_strat` stays clear of setoid rewriting.

But it turns out that `rewrite_strat` doesn't seem to work at all without `Proper` instances. Here's my (slightly modified) test case:

```coq
From Coq Require Import ZArith.
Import Relations RelationClasses Morphisms.               (* For smaller traces *)

Axiom mod_le : forall (a b : Z) (_ : (Z.le Z0 a))  (_ : (Z.lt a b)), @eq Z (Z.modulo a b) a.

Axiom trim : N -> Z -> Z.

Create HintDb no_proper discriminated.

Goal forall x,
  Z.le Z0
    (Z.add
       (Z.quot
          (Z.min (Z.modulo (Zpos (xO (xI (xI (xI (xO (xO xH))))))) (Z.pow (Zpos (xO xH)) (Zpos (xO (xO (xO (xO (xO (xO xH)))))))))
             (trim (Npos (xO (xO (xO (xO (xO (xO xH)))))))
                (Z.sub x (Z.modulo (Zpos (xO (xO (xI xH)))) (Z.pow (Zpos (xO xH)) (Zpos (xO (xO (xO (xO (xO (xO xH))))))))))))
          (Z.modulo (Zpos (xO xH)) (Z.pow (Zpos (xO xH)) (Zpos (xO (xO (xO (xO (xO (xO xH)))))))))) (Zpos xH)).
  intros x.
  Set Typeclasses Debug.
  rewrite_strat (topdown mod_le) with no_proper.
```

Unfortunately this fails on the branch of this MR with the following error message:

```
setoid rewrite failed: Unable to satisfy the following constraints:
UNDEFINED EVARS:
 ?X277==[x |- relation Z] (internal placeholder) {?r}
 ?X278==[x (do_subrelation:=do_subrelation) |-
          @Proper (forall _ : Z, Z) (@respectful Z Z (@eq Z) ?r) (Z.sub x)]
          (internal placeholder) {?p}
 ?X332==[x |- relation Z] (internal placeholder) {?r0}
 ?X333==[x (do_subrelation:=do_subrelation) |-
          @Proper (forall _ : Z, Z) (@respectful Z Z ?r ?r0)
            (trim (Npos (xO (xO (xO (xO (xO (xO xH))))))))]
          (internal placeholder) {?p0}
 ?X463==[x |- relation Z] (internal placeholder) {?r1}
 ?X464==[x (do_subrelation:=do_subrelation) |-
          @Proper (forall (_ : Z) (_ : Z), Z)
            (@respectful Z (forall _ : Z, Z) (@eq Z)
               (@respectful Z Z ?r0 ?r1)) Z.min] (internal placeholder) {?p1}
 ?X696==[x |- relation Z] (internal placeholder) {?r2}
 ?X697==[x (do_subrelation:=do_subrelation) |-
          @Proper (forall (_ : Z) (_ : Z), Z)
            (@respectful Z (forall _ : Z, Z) ?r1
               (@respectful Z Z (@eq Z) ?r2)) Z.quot] (internal placeholder) {?p2}
 ?X947==[x |- relation Z] (internal placeholder) {?r3}
 ?X948==[x |- relation Z] (internal placeholder) {?r4}
 ?X949==[x (do_subrelation:=do_subrelation) |-
          @Proper (forall (_ : Z) (_ : Z), Z)
            (@respectful Z (forall _ : Z, Z) ?r2 (@respectful Z Z ?r4 ?r3))
            Z.add] (internal placeholder) {?p3}
 ?X950==[x |- @ProperProxy Z ?r4 (Zpos xH)] (internal placeholder) {?p4}
 ?X1209==[x (do_subrelation:=do_subrelation) |-
           @Proper (forall _ : Z, Prop)
             (@respectful Z Prop ?r3
                (@Basics.flip Prop Prop Prop Basics.impl)) 
             (Z.le Z0)] (internal placeholder) {?p5}
TYPECLASSES:?X277 ?X278 ?X332 ?X333 ?X463 ?X464 ?X696 ?X697 ?X947 ?X948 ?X949
?X950 ?X1209
SHELF:||
FUTURE GOALS STACK:?X1209 ?X950 ?X949 ?X948 ?X947 ?X697 ?X696 ?X673 ?X672
?X464 ?X463 ?X333 ?X332 ?X278 ?X277 ?X236 ?X235 ?X76 ?X75||
```

I am not aware that my example actually requires setoid rewriting but I could be wrong.

**Comment by SkySkimmer** [comment]
rewrite_strat is setoid rewriting so IMO makes sense that it doesn't work without any instances
Maybe we can get a db with only the instances need for `Proper eq` (and associated `respectful` etc)?

**Comment by SkySkimmer** [comment]
>I was hoping that it could be avoided by using rewrite_strat .. with <empty_database> to make sure rewrite_strat stays clear of setoid rewriting.

rewrite_strat is setoid rewriting so there's no chance of that.

**Comment by SkySkimmer** [comment]
Doesn't seem useful enough to try to get it merged.

**Pull request closed without merging**